storage/remote/azuread/azuread.go

```diff
@@ -245,3 +317,17 @@ func getAudience(cloud string) (string, error) {
 		return "", errors.New("Cloud is not specified or is incorrect: " + cloud)
 	}
 }
+
+// getCloudConfiguration returns the cloud Configuration which contains AAD endpoint for different clouds
+func getCloudConfiguration(c string) (cloud.Configuration, error) {
```

Comment:
I'm a little confused on the functionalities of this function and the very similar `getAudience`, and why both are needed. I have an admittedly rough understanding of what's going on here, but when the token is requested here:

https://github.com/prometheus/prometheus/blob/2b2cfd1f89ec44d4012c388aa6ddeccc136f579f/storage/remote/azuread/azuread.go#L279

it uses a credential client with a client option for that cloud and an separate objects object with an audience scope for that same cloud. So I guess I why the cloud needed to be both specified in the client configuration and the request configuration?

Reply:
The reason I need to split this into different function primarily is because the source of truth for the values returned by these functions are different and also one is used at the time of creation of credential (cloudConfiguration) and the other is used at the time of creation of token provider (audience).

Coming to cloudConfiguration, this has the ActiveDirectoryAuthorityHost for different Azure clouds and we get this value from `github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud` package. This host value is required by the credential while trying to reach AzureAD. The values for ActiveDirectoryAuthorityHost are:
1. AzurePublic: https://login.microsoftonline.com/ 
2. AzureGovernment: https://login.microsoftonline.us/
3. AzureChina: https://login.chinacloudapi.cn/
If the cloudConfiguration is not passed to the credential, it by default picks AzurePublic ActiveDirectoryAuthorityHost

Now for the audience, the values for this is currently defined statically in the code and is again different for different Azure clouds. The value of the audience(or also commonly called as scope) is used to get a token with specific roles from AzureAD and hence, required during the initialization of the token provider. 

I hope this answers your question. If something is still not clear, feel free to reach out to me.

Reply:
Thanks for the explanation! I think I understand now.